q2_quality_filter/_filter.py

```diff
@@ -6,181 +6,448 @@
+        if trunc_fraction <= min_length_fraction:
+            status = RecordStatus.SHORT
+
+            return fastq_record, status
+
+    # mark ambiguous if too many ambiguous bases are present
+    if fastq_record.sequence.count(b'N') > max_ambiguous:
+        if status == RecordStatus.TRUNCATED:
+            status = RecordStatus.TRUNCATED_AMBIGUOUS
+        else:
+            status = RecordStatus.AMBIGUOUS
+
+    return fastq_record, status
+
+
+def _is_retained(
+    forward_status: RecordStatus,
+    reverse_status: RecordStatus | None,
+    filtering_stats_df: pd.DataFrame,
+    sample_id: str
+) -> bool:
+    '''
+    Determines whether a fastq record or pair of fastq records will retained
+    in the output. The `reverse_status` is None in the case of single-end
+    reads.
+
+    Parameters
+    ----------
+    forward_status : RecordStatus
+        The status of the record from the forward fastq file.
+    reverse_status : RecordStatus or None
+        The status of the record from the reverse fastq file if it exists
+        otherwise None.
+    filtering_stats_df : pd.DataFrame
+        The data structure that tracks filtering stats.
+    sample_id : str
+        The sample id that the record(s) belongs to.
+
+    Returns
+    -------
+    bool
+        True if the record(s) is to be retained, False otherwise.
+    '''
+    filtering_stats_df.loc[sample_id, 'total-input-reads'] += 1
+
+    if (RecordStatus.SHORT in (forward_status, reverse_status)):
+        filtering_stats_df.loc[sample_id, 'reads-truncated'] += 1
+        filtering_stats_df.loc[
+            sample_id, 'reads-too-short-after-truncation'
+        ] += 1
+        return False
+
+    if (RecordStatus.AMBIGUOUS in (forward_status, reverse_status)):
+        filtering_stats_df.loc[
+            sample_id, 'reads-exceeding-maximum-ambiguous-bases'
+        ] += 1
+        return False
+
+    if (RecordStatus.TRUNCATED_AMBIGUOUS in (forward_status, reverse_status)):
+        filtering_stats_df.loc[sample_id, 'reads-truncated'] += 1
+        filtering_stats_df.loc[
+            sample_id, 'reads-exceeding-maximum-ambiguous-bases'
+        ] += 1
+        return False
+
+    if (RecordStatus.TRUNCATED in (forward_status, reverse_status)):
+        filtering_stats_df.loc[sample_id, 'reads-truncated'] += 1
+
+    filtering_stats_df.loc[sample_id, 'total-retained-reads'] += 1
+
+    return True
+
+
+def _write_record(fastq_record: FastqRecord, fh: gzip.GzipFile) -> None:
+    '''
+    Writes a fastq record to an open fastq file.
+
+    Parameters
+    ----------
+    fastq_record : FastqRecord
+        The fastq record to be written.
+    fh : GzipFile
+        The output fastq file handler.
+
+    Returns
+    -------
+    None
+    '''
+    fh.write(fastq_record.sequence_header + b'\n')
+    fh.write(fastq_record.sequence + b'\n')
+    fh.write(fastq_record.quality_header + b'\n')
+    fh.write(fastq_record.quality_scores + b'\n')
+
+
+def q_score(
+    demux: Union[
```

Comment:
I think this `Union` is `_ReadDirectionUnion`.

Reply:
true, we could substitute that if you prefer

Reply:
Might as well, since you've defined it.

Reply:
Actually this isn't possible because `_ReadDirectionUnion` only goes to single/paired end and not the other way around.